lib/src/tls.rs

```diff
@@ -0,0 +1,1038 @@
+//! # Tls module
+//!
+//! This module p certificate: (), certificate_chain: (), key: (), versions: ()  certificate: (), certificate_chain: (), key: (), versions: () rovides traits and structures to handle tls. It provides a unified
+//! certificate resolver for rustls and openssl.
+use std::{
+    borrow::ToOwned,
+    collections::{HashMap, HashSet},
+    convert::From,
+    io::BufReader,
+    sync::{Arc, Mutex},
+};
+
+use crate::router::trie::*;
+use crate::sozu_command::proxy::{
+    AddCertificate, CertificateAndKey, CertificateFingerprint, RemoveCertificate,
+    ReplaceCertificate,
+};
+
+use rustls::{
+    internal::pemfile,
+    sign::{CertifiedKey, RSASigningKey},
+    ClientHello, ResolvesServerCert,
+};
+use sha2::{Digest, Sha256};
+use sozu_command::proxy::TlsVersion;
+use x509_parser::{
+    oid_registry::{OID_X509_COMMON_NAME, OID_X509_EXT_SUBJECT_ALT_NAME},
+    pem::{parse_x509_pem, Pem},
+};
+
+// -----------------------------------------------------------------------------
+// CertificateResolver trait
+
+pub trait CertificateResolver {
+    type Error;
+
+    // `get_certificate` return the certificate, its chain and private key
+    fn get_certificate(
+        &self,
+        fingerprint: &CertificateFingerprint,
+    ) -> Option<ParsedCertificateAndKey>;
+
+    // `add_certificate` to the certificate resolver, ensure that it is valid and check if it can
```

Comment:
```suggestion
    // adds a certificate to the certificate resolver, ensures that it is valid and checks if it can
```

Reply:
I will do not update the beginning of the sentence to match the name of the function